lib/resty/wasmx/shm.lua

```diff
@@ -149,144 +155,163 @@ local _metric_type_set = {
-local _mbuf = ffi_new("u_char [?]", _mbs)
-local _hbs = C.ngx_wa_ffi_shm_max_histogram_size()
-local _hbuf = ffi_new("u_char [?]", _hbs)
+local _mbs = C.ngx_wa_ffi_shm_metrics_one_slot_size()
+local _hbs = C.ngx_wa_ffi_shm_metrics_histogram_max_size()
+local _mbuf = ffi_new("u_char[?]", _mbs)
+local _hbuf = ffi_new("u_char[?]", _hbs)
+local _kbuf = ffi_new("ngx_str_t *[?]", DEFAULT_KEYS_PAGE_SIZE)
 
 
-local function lock_shm(zone)
+local function shm_lock(zone)
     C.ngx_wa_ffi_shm_lock(zone[WASM_SHM_KEY])
 end
 
 
-local function unlock_shm(zone)
+local function shm_unlock(zone)
     C.ngx_wa_ffi_shm_unlock(zone[WASM_SHM_KEY])
 end
 
 
 local function key_iterator(ctx)
-    if ctx.i < ctx.page_total then
-        local ckey = ctx.ckeys[ctx.i]
-        ctx.i = ctx.i + 1
+    if ctx.i == tonumber(ctx.ccur_index[0]) then
+        ngx_sleep(0) -- TODO: support non-yielding phases
 
-        return ffi_str(ckey.data, ckey.len)
+        ctx.i = 0
+        ctx.ccur_index[0] = 0
 
-    else
-        ctx.cpage_total[0] = 0
+        local rc = C.ngx_wa_ffi_shm_iterate_keys(ctx.shm, ctx.page_size,
+                                                 ctx.clast_index, ctx.ccur_index, ctx.ckeys)
 
-        local rc = C.ngx_wa_ffi_shm_iterate_keys(ctx.shm,
-                                                 ctx.page, ctx.page_size,
-                                                 ctx.ckeys, ctx.cpage_total)
         if rc == FFI_ABORT then
+            -- users must manage locking themselves (e.g. break condition in the for loop)
             local zone_name = ffi_str(ctx.shm.name.data, ctx.shm.name.len)
-            local err = "attempt to iterate over the keys of an unlocked shm zone."
-                .. " please call resty.wasmx.shm.%s:lock() before calling"
-                .. " iterate_keys() and resty.wasmx.shm.%s:unlock() after"
+            local err = "attempt to call %s:iterate_keys() but the " ..
+                        "shm zone is not locked; invoke :lock() before " ..
+                        "and :unlock() after."
 
-            error(str_fmt(err, zone_name, zone_name), 2)
+            error(str_fmt(err, zone_name), 2)
         end
 
-        if rc == FFI_DECLINED then
+        if rc == FFI_DONE then
             return
         end
 
-        ngx_sleep(0)
-
-        ctx.page_total = tonumber(ctx.cpage_total[0])
-        ctx.page = ctx.page + ctx.page_total
-        ctx.i = 1
+        if rc ~= FFI_OK then
+            error("unknown error")
+        end
 
-        return ffi_str(ctx.ckeys[0].data, ctx.ckeys[0].len)
+        ngx_log(DEBUG, "iterate_keys fetched a new page")
     end
+
+    local ckey = ctx.ckeys[ctx.i]
+    local key = ffi_str(ctx.ckeys[ctx.i].data, ctx.ckeys[ctx.i].len)
+
+    ctx.i = ctx.i + 1
+
+    return key
 end
 
 
-local function iterate_keys(zone, page_size)
-    if page_size ~= nil and type(page_size) ~= "number" then
-        error("page_size must be a number", 2)
+local function shm_iterate_keys(zone, opts)
```

Comment:
I think grouping optional arguments into `opts` is an overkill here considering that `key_iterator` is likely to only be configured in two ways: page size and yield/not_yield.

I must admit, the additional complexity required for validating grouped optional arguments annoys me.

Reply:
Hm, `iterate_keys(1024, false)` does look more obscure. Fair point.